test/e2e_node/pod_ips.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	watchtools "k8s.io/client-go/tools/watch"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	"k8s.io/kubernetes/test/e2e/network/common"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+var _ = common.SIGDescribe("Pod IPs", framework.WithSerial(), func() {
+	f := framework.NewDefaultFramework("pod-ips")
+	f.NamespacePodSecurityLevel = admissionapi.LevelRestricted
+
+	ginkgo.Context("when pod gets terminated", func() {
+		ginkgo.It("should contain podIPs in status for terminal pod", func(ctx context.Context) {
+			podName := "pod-ips-" + string(uuid.NewUUID())
+
+			podSpec := e2epod.MustMixinRestrictedPodSecurity(&v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: podName,
+				},
+				Spec: v1.PodSpec{
```

Comment:
can you please add two more tests:

1. Let the container crash instead of exiting normally. Variations may be oom kill, error in PostStart hook, and never even started because of the image pull problem. But simple crash will likely suffice.
2. Ensure PodIP is available while container is terminating (executing preStop hook). This is not affected by your change, but nice to test anyways for future regressions

Reply:
I added two new test cases to hopefully capture the intention. They all fail on `master` and pass on the branch:

> * Let the container crash instead of exiting normally. Variations may be oom kill, error in PostStart hook, and never even started because of the image pull problem. But simple crash will likely suffice.

Here I simulate the crash by simply returning a non-zero exit code. 

I tested manually OOM (it fails on master, and works ok on the branch), and image pull issue (works ok both on master and the branch, a deleted pod transitions directly from Pending to Failed in that case).

> * Ensure PodIP is available while container is terminating (executing preStop hook). This is not affected by your change, but nice to test anyways for future regressions

Here I delete the pod to trigger the termination, and use a finalizer to be sure I can inspect the status.